src/action/list_directory.rs

```diff
@@ -0,0 +1,689 @@
+        path: file_path.clone().to_path_buf(),
+        crtime: get_creation_time(&metadata),
+    }
+}
+
+#[cfg(not(target_os = "linux"))]
+fn fill_response(metadata: &Metadata, file_path: &Path) -> Response {
+    Response {
+        size: Some(metadata.len()),
+        atime: get_accesses_time(&metadata),
+        mtime: get_modification_time(&metadata),
+        path: file_path.clone().to_path_buf(),
+        crtime: get_creation_time(&metadata),
+        ..Default::default()
+    }
+}
+
+pub fn handle<S: Session>(session: &mut S, request: Request)
+                          -> session::Result<()> {
+    let dir_path = &request.path;
+    let mut paths: Vec<PathBuf> = dir_path.read_dir()
+        .map_err(Error::ReadPath)?.filter_map(|entry| entry.ok())
+        .map(|entry| entry.path()).collect();
+    paths.sort();
+
+    for file_path in &paths {
+        let metadata = fs::symlink_metadata(file_path)
+            .map_err(Error::ReadPath)?;
+        session.reply(fill_response(&metadata, file_path))?;
+    }
+
+    Ok(())
+}
+
+fn get_path(path: &Option<String>) -> PathBuf {
+    match path {
+        Some(string_path) if !string_path.is_empty() => {
+            PathBuf::from(string_path)
+        }
+        _ => PathBuf::from("/"),
+    }
+}
+
+/// Fills st_linux_flags field
+#[cfg(target_os = "linux")]
+fn get_linux_flags(path: &Path) -> Option<c_long> {
+    use std::fs::File;
+    use std::os::unix::io::AsRawFd;
+
+    let file = match File::open(path) {
+        Ok(file) => file,
+        Err(_) => return None,
+    };
+    let mut linux_flags: c_long = 0;
+    let linux_flags_ptr: *mut c_long = &mut linux_flags;
+    unsafe {
+        match ioctls::fs_ioc_getflags(file.as_raw_fd(), linux_flags_ptr) {
+            0 => Some(linux_flags),
+            _ => None,
+        }
+    }
+}
+
+impl super::Request for Request {
+
+    type Proto = ListDirRequest;
+
+    fn from_proto(proto: Self::Proto) -> Result<Request, session::ParseError> {
+        let missing = session::MissingFieldError::new;
+        let pathspec = proto.pathspec.ok_or(missing("path spec"))?;
+        let path_type = pathspec.pathtype
+            .ok_or(missing("path type"))?;
+        if path_type != PathType::Os as i32 {
+            return Err(unsupported_value(String::from("path type"),
+                                         path_type.to_string()));
+        }
+        let path_option = pathspec.path_options
+            .unwrap_or(Options::CaseLiteral as i32);
+        if path_option != Options::CaseLiteral as i32 {
+            return Err(unsupported_value(String::from("path option"),
+                                         path_option.to_string()));
+        };
+        Ok(Request {
+            path: get_path(&pathspec.path),
+        })
+    }
+}
+
+fn get_time_since_unix_epoch(sys_time: &Option<SystemTime>) -> Option<u64> {
+    return sys_time.map_or(None, |time| time.duration_since(UNIX_EPOCH)
+        .map_or(None, |dur| Some(dur.as_secs())));
+}
```

Comment:
I think you can use [this](https://github.com/google/rrg/blob/master/proto/src/lib.rs#L218-L221).

Reply:
Done, thx.